activerecord/lib/active_record/railties/databases.rake

```diff
@@ -166,11 +166,26 @@ db_namespace = namespace :db do
 
   desc "Rolls the schema back to the previous version (specify steps w/ STEP=n)."
   task rollback: :load_config do
-    step = ENV["STEP"] ? ENV["STEP"].to_i : 1
-    ActiveRecord::Base.connection.migration_context.rollback(step)
+    ActiveRecord::Base.configurations.configs_for(env_name: Rails.env).each do |db_config|
```

Comment:
I would expect `db:rollback` to roll back the most recent migration across all databases, not to roll back the most recent migration on each database.

Reply:
I'm not sure that's correct. When you run a migration it only runs on the current environment and then dumps the structure into the test env.

Reply:
By "across all databases", I mean the set of databases for a single environment.

To give a concrete example: if I have a `primary` database that contains a `users` table and an `animals` database that contains a `dogs` table, and I generate and run a migration that adds a `badgers` table to the `animals` database, I would expect `db:rollback` to only roll back the migration that added the `badgers` table, and to not also roll back the migration on the `primary` database that created the `users` table, which is what is implemented here.

Reply:
I like this idea, but it sounds considerably more complex to implement. I'll give it a shot in another branch.

Reply:
Particularly complicated by the fact we can't rely on all the DBs using the same numbering scheme for their migrations; timestamps are just a default / suggestion. #21670 comes to mind, but I don't think we have an answer for [the structure.sql issue](https://github.com/rails/rails/pull/21670#issuecomment-168336331).

Reply:
Here's a naive approach https://github.com/rails/rails/compare/master...gmcgibbon:db_rollback_multi_db_2. It seems kind of weird that migrations wouldn't be either all numbers or all timestamps, but that's definitely something we should think about.

Should we raise an exception for that case, or add `created_at` to schema migrations as that PR suggests?

Reply:
Even if they're all integers, they're not _comparable_: the right thing to roll back may be `db/animals_migrate/3_add_dogs`, irrespective of the 500 files that have been sitting in `db/migrate` for years.

Reply:
I agree that the behavior I expect is what @eugeneius described but I don't know how to solve that other than adding a timestamp to the migrations table. The problem with that is the structure.sql file and we will need a way to transform the SQL dump of each datastore that seems complicated to me.

Reply:
Actually.. shouldn't the SQL dump ignore the migrations table entirely, as the schema file does?

Reply:
On a trivial database, I get:
```sql
CREATE TABLE IF NOT EXISTS "schema_migrations" ("version" varchar NOT NULL PRIMARY KEY);
CREATE TABLE IF NOT EXISTS "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL);
CREATE TABLE sqlite_sequence(name,seq);
CREATE TABLE IF NOT EXISTS "routes" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL, "shard" integer NOT NULL, "blog_id" integer NOT NULL);
CREATE INDEX "index_routes_on_blog_id" ON "routes" ("blog_id");
INSERT INTO "schema_migrations" (version) VALUES
('20180915214708');
```

So both `schema_migrations ` and `ar_internal_metadata` are tracked. Maybe we should ignore them seeing how the ruby schema format doesn't track them? `ActiveRecord::SchemaDumper.ignore_tables` is empty by default.